lib/autoload/polyglot/language.rb

```
@@ -0,0 +1,92 @@
+# instance for the language. When defining a new language, add it to the list of autoloaded classes.
+class Polyglot::Language < ActiveRecord::Base
+  extend ActiveSupport::Autoload
+
+  self.table_name = 'polyglot_languages'
+  acts_as_forest
+  eager_autoload do
+    autoload :Python
+  end
+
+  after_initialize :set_readonly
+  after_save :set_readonly
+
+  validate :unique_root_language, unless: :parent
+
+  # @!method self.instance
+  #   Gets an instance of the language object. Any class which implements this method can be
+  #   instantiated and is expected to no longer be abstract.
+  #
+  #   @abstract
+  #   @return [Polyglot::Language]
+
+  # Gets the display name of the language.
+  #
+  # @abstract
+  # @return [String]
+  def self.display_name
+    fail NotImplementedError
+  end
+
+  # The stylesheets that need to be packaged with the rest of the application.
+  #
+  # This should include the Rouge/Pygments stylesheet for formatting code.
+  #
+  # @abstract
+  # @return [Array<String>]
+  def self.stylesheets
+    fail NotImplementedError
+  end
+
+  # The script files that need to be packaged with the rest of the application.
+  #
+  # This should include the Ace mode for the language.
+  #
+  # @abstract
+  # @return [Array<String>]
+  def self.javascript
+    fail NotImplementedError
+  end
+
+  # Finds or creates the root instance for languages of this class.
+  #
+  # @return [Polyglot::Language]
+  def self.root
+    # Creating the root record would call +root+, so if we are defining @root, we return nil.
+    if instance_variable_defined?(:@root)
+      @root
+    else
+      @root = nil
+      @root = find_or_create_by(name: display_name, parent: nil)
+    end
+  end
+  private_class_method :root
+
+  # Loads or creates the database records for each defined programming language.
+  def self.load_languages
+    descendants.each do |klass|
+      klass.instance if klass.respond_to?(:instance)
+    end
+  rescue ActiveRecord::StatementInvalid
```

Do not suppress exceptions.